Draft 1: Kubernetes documentation (#3115)
mattermost/docs: source/install/use-kubernetes-mattermost.rst

.. _use-kubernetes-mattermost:

Using Mattermost Operator Functionality
======================================

Rolling Upgrades
----------------

The Mattermost Kubernetes Operator supports rolling upgrades so you can upgrade
your Mattermost deployment with zero downtime. This process
requires at least two replicas as a rolling upgrade cannot be performed if there is only one pod.

> Replicas are created when a user count is selected and exceeds 100.

Reviewer: Does this happen automatically?

Author: Do you mean the rolling upgrades or the creation of the replicas?

Reviewer 2: This line is clarifying that rolling upgrades are only possibly for deployments of more than 100 users right now since you need more than 1 replica

> New releases are announced via our community server, as well as social media.

Reviewer: This line seems out of place to me. Is this for the MM operator itself or for the Rolling upgrade feature?

> New Mattermost releases are announced via our community server, as well as social media.

Reviewer 2: ```suggestion
New Mattermost releases are announced via our community server, as well as social media and email.
```

**To perform a rolling upgrade:**

1. Log in to your Kubernetes instance.
2. Open the ``clusterinstallation.yaml`` manifest (the one created during installation).
3. Update the ``spec.version`` value to the new version.
4. Save the changes.

Apply the changes with kubectl using

.. code-block:: sh

  $ kubectl apply -n mattermost -f /path/to/cluster-installation.yaml

> The operator initiates a job to perform the database migration (if applicable).

Reviewer 2: It will always create the job but the job will only perform a DB migration if one is needed

> The job is created in the Kubernetes cluster and once migration is complete the pods are restarted.

Reviewer 2: ```suggestion
The job is created in the Kubernetes cluster and once migration is complete the pods are rolling restarted with the new version.
```

To view information about the running job, use

.. code-block:: sh

  $ kubectl -n mattermost get jobs

At least one pod is available at all times and once all pods are restarted with the new version the upgrade is complete.

To view the status of the pods and to confirm their state, use

.. code-block:: sh

  $ kubectl -n mattermost get pods

> The *STATUS* of the pods should be running/ready, with an *AGE* of a few seconds.

Reviewer 2: ```suggestion
The *STATUS* of the pods should be running/ready, with an *AGE* of 10-15 seconds.
```

Blue-green Deployments
----------------------

Blue-green deployments can reduce downtime and increase stability during automated tasks in a production environment.
This technique entails running two identical production environments in tandem, with one acting as a
live environment and one as an idle environment.

An upgrade can be rolled out to the idle environment (“blue”) while the other (“green”) is the production environment.
Once the upgrade has been successfully rolled out, all traffic can be switched from “green” to “blue”.

> **Note:** Blue/green can be run on a permanent basis but utilizes more resources in the Kubernetes cluster than normal deployments.

Reviewer 2: ```suggestion
**Note:** Blue-green can be run on a permanent basis but utilizes more resources in the Kubernetes cluster than normal deployments.
```
